hibernate-envers/src/main/java/org/hibernate/envers/query/criteria/internal/EntityTypeAuditExpression.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.envers.query.criteria.internal;
+
+import java.util.Map;
+
+import org.hibernate.envers.boot.internal.EnversService;
+import org.hibernate.envers.internal.reader.AuditReaderImplementor;
+import org.hibernate.envers.internal.tools.query.Parameters;
+import org.hibernate.envers.internal.tools.query.QueryBuilder;
+import org.hibernate.envers.query.criteria.AuditCriterion;
+
+/**
+ * @author Felix Feisst (feisst dot felix at gmail dot com)
+ */
+public class EntityTypeAuditExpression implements AuditCriterion {
+
+	private String alias;
+	private String entityName;
+
+	public EntityTypeAuditExpression(
+			String alias,
+			String entityName) {
+		this.alias = alias;
+		this.entityName = entityName;
+	}
+
+	@Override
+	public void addToQuery(
+			EnversService enversService,
+			AuditReaderImplementor auditReader,
+			Map<String, String> aliasToEntityNameMap,
+			String baseAlias,
+			QueryBuilder qb,
+			Parameters parameters) {
+		String effectiveAlias = alias == null ? baseAlias : alias;
+		String effectiveEntityName = entityName;
+		if ( enversService.getEntitiesConfigurations().isVersioned( effectiveEntityName ) ) {
```

Comment:
That's true.

I do believe we should look at changes to `AuditCriterion#addToQuery` to be provided more contextual information on the current state of the query generation process.  That would for when `EntityC` (not-audited) extends `EntityA` (audited), we could throw an exception but still be lenient and not do that for non-audited relation traversals.

I'll make a note in the code on merge indicating this unless you feel strongly otherwise.